File: src/main/java/boat/utilities/PropertiesHelper.java
package boat.utilities;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.Properties;

public class PropertiesHelper {


    public static final String PROPERTIY_FILE = "boat.cfg";
    public static final String PROPERTIY_FILE_DEFAULT = "boat.default.cfg";
    public static final String VERSION_FILE = "version.properties";

    public static String getVersion() {
        InputStream inputStream;
        String result = null;

        try {

            Properties prop = new Properties();

            inputStream = PropertiesHelper.class.getClassLoader().getResourceAsStream(VERSION_FILE);

            if (inputStream != null) {
                prop.load(inputStream);
            } else {
                throw new FileNotFoundException("property file '" + VERSION_FILE + "' not found in the classpath");
            }

            result = prop.getProperty("version");
            inputStream.close();

        } catch (Exception e) {
            System.out.println("Exception: " + e);
        }

        return result;
    }

    public static String getProperty(String propname) {
        InputStream inputStream;
        String result = null;

        // check if property exists in Environment, then skip
        String envValue = System.getenv(propname);
        if(envValue != null){
            return envValue;
        }

        if (getPropertyFile() != null) {
            try {

                Properties prop = new Properties();
                inputStream = new FileInputStream(getPropertyFile());

                if (inputStream != null) {
                    prop.load(inputStream);
                } else {
                    throw new FileNotFoundException("property file '" + getPropertyFile() + "' not found in the classpath");
                }

                result = prop.getProperty(propname);
                inputStream.close();

            } catch (Exception e) {
                System.out.println("Exception: " + e);
            }
        }
        return result;
    }

    private static String getPropertyFile() {
        if (new File(PROPERTIY_FILE).isFile()) {
            return PROPERTIY_FILE;
        } else if (new File(PROPERTIY_FILE_DEFAULT).isFile()) {
            return PROPERTIY_FILE_DEFAULT;
        }
        return null;
    }

}

File: src/main/java/boat/utilities/PropertiesHelper.kt
package boat.utilities

import java.io.File
import java.io.FileInputStream
import java.util.*

object PropertiesHelper {
    const val PROPERTIY_FILE: String = "boat.cfg"
    const val PROPERTIY_FILE_DEFAULT: String = "boat.default.cfg"
    const val VERSION_FILE: String = "version.properties"

    val version: String?
        get() = runCatching {
            PropertiesHelper::class.java.classLoader.getResourceAsStream(VERSION_FILE)?.use { inputStream ->
                Properties().apply { load(inputStream) }.getProperty("version")
            } ?: "version-missing"
        }.onFailure { println("Exception: $it") }.getOrNull()

    fun getProperty(propname: String, default: String? = null): String? {
        System.getenv(propname)?.let { return it }

        return propertyFile?.let { path ->
            runCatching {
                FileInputStream(path).use { inputStream ->
                    Properties().apply { load(inputStream) }.getProperty(propname)
                }
            }.onFailure { println("'$propname' Notfound, using :'$default'") }.getOrElse({ default })
        }
    }

    private val propertyFile: String?
        get() = when {
            File(PROPERTIY_FILE).isFile -> PROPERTIY_FILE
            File(PROPERTIY_FILE_DEFAULT).isFile -> PROPERTIY_FILE_DEFAULT
            else -> null
        }
}
